### PR TITLE
[air - preprocessor] Add BatchMapper.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

Add BatchMapper preprocessor.
Update the semantics of preprocessor.fit() to allow for multiple fit. This is to follow scikitlearn example.
Introduce FitStatus to explicitly incorporate Chain case.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/preprocessor.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:102]: Hmmm, got it. We can decide on the naming. But semantics are basically:
1. `fittable` is an inherent attribute of the "type" of the preprocessors. It also implies whether a `fit` method in meaningful at all throughout the entire lifetime of this preprocessor. 
2. should_fit/can_fit depends on the state a preprocessor is currently in (assuming it's fittable).

Exposing `check_is_fitted` alone is not enough, as you can see in `trainer.py` - it only checks for `check_is_fitted` in current impl, which leads to crash in the case of non-fittable preprocessors. That's why the proposal is to add `should_fit`.

`check_is_fitted` v.s. `is_fitted` or `can` v.s. `should` - I don't have much preference. @clarkzinzow @matthewdeng maybe as original author of the API?

Comment (gjoliver) [reply python/ray/ml/preprocessor.py:102]: agree with the change. I am just nit-picking the naming.
hope to get things named consistently. thanks :)

Comment (clarkzinzow) [reply python/ray/ml/preprocessor.py:102]: I opted for `should_fit` over `can_fit` since it's not indicating an optional operation for fittable preprocessors, it's a necessary operation: if a fittable preprocessor is not fit before calling `.transform()`, it will fail. An argument could be made for `need_fit`.

File: python/ray/ml/preprocessor.py
Comment (clarkzinzow) [inline python/ray/ml/preprocessor.py:48]: Could also make this a no-op when `not self._is_fittable`, which would be more friendly to e.g. chain preprocessors.

**Review by clarkzinzow — COMMENTED** [review]
Looking good, the biggest remaining things are:
1. We need to modify `Chain` to work correctly with the new `should_fit()` API.
2. I think that `BatchMapper.fit()` should be a no-op in order for `Chain` to be able to naively call `.fit()` and `.fit_transform()` on all of its preprocessors, which should be cleaner.

https://github.com/ray-project/ray/blob/8b8afd5b24f90e6fc6e3279849e0273b24718948/python/ray/ml/preprocessors/chain.py#L22-L47

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [reply python/ray/ml/preprocessor.py:102]: WE can change `check_is_fitted` to `is_fitted`.

`should_fit` functionally is a bit strange to me, at least as a public API. In particular, I want to avoid the case where the user does something like:

```python
if (preprocessor.should_fit()):
    preprocessor.fit()
```

It's not clear how to differentiate the case where the preprocessor is fitted from the case where the preprocessor was already fitted before.

Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:102]: @matthewdeng hmm, mind elaborating a bit more?
so 3 cases:
1. not fittable
2. fittable and fitted
3. fittable and not fitted yet

should_fit == case 3
if_fitted == 2 conditioned on (2 + 3)

Another way that I can see to work is just to enforce "at most once" fitting semantics internally - and caller doesn't have to call `should_fit` before `fit`. Which one do you prefer? Or are you proposing an alternative?

File: python/ray/ml/preprocessor.py
Comment (gjoliver) [inline python/ray/ml/preprocessor.py:67]: this error message looks weird. why don't you check:
assert self.should_fit() here as well?

File: python/ray/ml/preprocessor.py
Comment (gjoliver) [inline python/ray/ml/preprocessor.py:83]: should we update this error msg to be:

fit must be called before transform, or simply use fit_transform() to run both steps.

File: python/ray/ml/preprocessor.py
Comment (gjoliver) [reply python/ray/ml/preprocessor.py:102]: I think I lack the context why you want to bundle these 2 things together in the first place.
but in my mind, the most intuitive way is to:

if calling fit() or fit_transform(), and not fittable, throw exception.
if calling fit(), and already fitted, print warning msg, and no-op.
if calling fit_transoform(), and already fitted, print warning msg, then proceed to transform().

Comment (matthewdeng) [reply python/ray/ml/preprocessor.py:102]: Can we raise an exception in the `fit()/fit_transfomr()` when already fitted instead? Logging is better than no logging, but I worry the behavior here isn't clear for users (I can see users thinking it should re-fit).

**Comment by xwjiang2010** [comment]
@gjoliver @matthewdeng @clarkzinzow 
A few updates:
- introduces a `FitStatus` and `fit_status()` to incorporate some of the nuances for chained preprocessors.
- explicit exceptions thrown
- check_is_fitted is now private

File: python/ray/ml/preprocessors/chain.py
Comment (clarkzinzow) [reply python/ray/ml/preprocessors/chain.py:22]: I think you could support chain of chains by having a `partially_fitted_count` that you increment when such a partially fitted chain is encountered, then your return stats logic changes a bit:
```python
if fittable_count == 0:
    return Preprocessor.FitStatus.NON_FITTABLE
elif fitted_count == fittable_count:
    return Preprocessor.FitStatus.FITTED
elif fitted_count > 0 or partially_fitted_count > 0:
    return Preprocessor.FitStatus.PARTIALLY_FITTED
else:
    return Preprocessor.FitStatus.NOT_FITTED
```

File: python/ray/ml/trainer.py
Comment (xwjiang2010) [reply python/ray/ml/trainer.py:195]: See @matthewdeng's preference about wanting explicit exception.  :) 
let's make a decision and stick to it.

File: python/ray/ml/preprocessors/batch_mapper.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessors/batch_mapper.py:27]: Good catch! Removed.

File: python/ray/ml/preprocessors/chain.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessors/chain.py:22]: Yeah I think it could potentially work. Though I am a fan of not over generalizing things in favor of clarity and robustness of code. Happy to enhance the logic if we really have a case here.

File: python/ray/ml/preprocessor.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:36]: Sounds good :)

File: python/ray/ml/preprocessors/chain.py
Comment (gjoliver) [reply python/ray/ml/preprocessors/chain.py:22]: it's fine to start with this.
but probably the comment needs to be updated so it is not from your point of view?? :)

```
# AIR does not support a chain of chained preprocessors at this point.
# Assert to explicitly call this out.
# This can be revisited if compelling use cases emerge.
```

File: python/ray/ml/trainer.py
Comment (gjoliver) [reply python/ray/ml/trainer.py:195]: I think we should allow fitted dataset, and basically no-op here.
why do we want to require unfitted dataset? what if the entire dataset is not_fitable?

Comment (xwjiang2010) [reply python/ray/ml/trainer.py:195]: we could do that. It's just @matthewdeng has this concern to not silently no-op (even with a warning msg):
```
Can we raise an exception in the fit()/fit_transfomr() when already fitted instead? Logging is better than no logging, but I worry the behavior here isn't clear for users (I can see users thinking it should re-fit).
```

Comment (gjoliver) [reply python/ray/ml/trainer.py:195]: print a info or warning msg sounds good.

Comment (clarkzinzow) [reply python/ray/ml/trainer.py:195]: So I think that `Preprocessor` itself should error if `.fit()` is called on an already fitted preprocessor, but I was less sure about whether Train as a user of `Preprocessor` should let these exceptions happen. I think that @matthewdeng is right, we should error here to ensure that the user doesn't think that an overwriting or incremental fit is happening.

Comment (gjoliver) [reply python/ray/ml/trainer.py:195]: what about partially fitted chain? what's a user's options here?

Comment (xwjiang2010) [reply python/ray/ml/trainer.py:195]: Synced offline.
@matthewdeng @gjoliver @clarkzinzow PTAL.

File: python/ray/ml/preprocessor.py
Comment (clarkzinzow) [inline python/ray/ml/preprocessor.py:71]: Do we want to use `warnings` or `logger.warning`? Not sure what the standard is for the ML libs, but core/data tends to use the latter.

File: python/ray/ml/preprocessor.py
Comment (clarkzinzow) [inline python/ray/ml/preprocessor.py:113]: Nit:
```suggestion
                "`fit` must be called before `transform`, "
                "or simply use fit_transform() to run both steps."
```

**Review by clarkzinzow — APPROVED** [review]
LGTM, only nits! IMO good to merge after one other ML team reviewer approval.

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:34]: Name these the same?

```suggestion
        PARTIALLY_FITTED = "PARTIALLY_FITTED"
```

File: python/ray/ml/preprocessors/chain.py
Comment (matthewdeng) [inline python/ray/ml/preprocessors/chain.py:23]: Add message so the user knows the reason.

File: python/ray/ml/preprocessors/batch_mapper.py
Comment (matthewdeng) [inline python/ray/ml/preprocessors/batch_mapper.py:29]: nit: Does `self.fn` actually turn into a readable string?

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:136]: From other recent suggestions, it's probably easier to use this as a property:

```suggestion
    @property
    def _is_fitted(self) -> bool:
```

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:33]: Explain what this means for the chain case!

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:54]:
```suggestion
        Calling it more than once will replace previously fitted state: ``preprocessor.fit(A).fit(B)`` is equivalent to ``preprocessor.fit(B)``.
```

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:64]: Don't need to be so specific to Chain here.
```suggestion
            # No-op because there is no state to be fitted.
```

File: python/ray/ml/preprocessor.py
Comment (matthewdeng) [inline python/ray/ml/preprocessor.py:74]:
```suggestion
                "`fit` has already been called on this Preprocessor (or on at least one contained Preprocessor if this is a Chain). All previously fitted state will be overwritten."
```

File: python/ray/ml/tests/test_preprocessors.py
Comment (matthewdeng) [inline python/ray/ml/tests/test_preprocessors.py:105]: Verify the results are that of the second dataset?

**Review by matthewdeng — COMMENTED** [review]
New functionality looks great, thanks for iterating on this!

File: python/ray/ml/preprocessor.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:71]: https://docs.python.org/3/howto/logging.html#when-to-use-logging

```
`warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning

`logging.warning()` if there is nothing the client application can do about the situation, but the event should still be noted
```

File: python/ray/ml/preprocessors/batch_mapper.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessors/batch_mapper.py:29]: it can look like: 
```
<BatchMapper udf=<function test_batch_mapper.<locals>.add_and_modify_udf at 0x7fb409149290>>
```

I don't think it is a big issue. What is the alternative?

File: python/ray/ml/preprocessor.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:136]: prefer to keep it this way so not to be confused with `fit_status()`, which is actually more acting like property in this case.

File: python/ray/ml/preprocessors/batch_mapper.py
Comment (matthewdeng) [reply python/ray/ml/preprocessors/batch_mapper.py:29]: Ah didn't realize it would come out like that. Seems okay for now, though we should probably change this (or implement `__str__`) for all preprocessors to something more concise since this will get included in the Tune Trial name.

Comment (clarkzinzow) [reply python/ray/ml/preprocessors/batch_mapper.py:29]: Could do 
```suggestion
        return f"<BatchMapper udf={self.fn.__name__}>"
```
A little bit nicer.

Comment (xwjiang2010) [reply python/ray/ml/preprocessors/batch_mapper.py:29]: updated.

File: python/ray/ml/preprocessor.py
Comment (xwjiang2010) [reply python/ray/ml/preprocessor.py:54]: updated. also the fit_transform method.

**Review by matthewdeng — COMMENTED** [review]
LGTM - can you update the PR summary and add a description (including the fit status changes)?

**Review by matthewdeng — APPROVED** [review]

**Pull request merged**
